k8s/src/main/scala/io/buoyant/k8s/EndpointsNamer.scala

```diff
@@ -53,59 +53,48 @@ class EndpointsNamer(idPrefix: Path, mkApi: String => NsApi) extends Namer {
   private[this] object Ns {
     private[this] var caches: Map[String, NsCache] = Map.empty
 
-    def get(name: String): NsCache = synchronized {
+    def get(name: String): Future[NsCache] = synchronized {
       caches.get(name) match {
-        case Some(ns) => ns
+        case Some(ns) =>
+          Future.value(ns)
         case None =>
-          val ns = mkNs(name)
-          caches += (name -> ns)
-          ns
+          mkNs(name).onSuccess { ns =>
+            caches += (name -> ns)
+          }
       }
```

Comment:
This is racey:

T0: caches.get(name) => None; mkNs(name)
T1: caches.get(name) => None; mkNs(name)

Furthermore, updates to the cache are done from another thread and are not synchronized.